File: frontend/src/lib/utils/markdown.ts
export type MarkdownSegment =
	| { type: 'text'; value: string }
	| { type: 'link'; text: string; href: string };

const LINK_RE = /\[([^\]]+)\]\(([^)\s]+)\)/g;
const SAFE_SCHEME = /^(https?:\/\/|mailto:|\/|#)/i;

export function parseMarkdownLinks(input: string): MarkdownSegment[] {
	if (!input) return [];
	const segments: MarkdownSegment[] = [];
	let lastIndex = 0;
	LINK_RE.lastIndex = 0;
	let match: RegExpExecArray | null;
	while ((match = LINK_RE.exec(input)) !== null) {
		const [whole, text, href] = match;
		if (!SAFE_SCHEME.test(href)) continue;
		if (match.index > lastIndex) {
			segments.push({ type: 'text', value: input.slice(lastIndex, match.index) });
		}
		segments.push({ type: 'link', text, href });
		lastIndex = match.index + whole.length;
	}
	if (lastIndex < input.length) {
		segments.push({ type: 'text', value: input.slice(lastIndex) });
	}
	return segments;
}

export function hasMarkdownLink(input: string): boolean {
	if (!input) return false;
	LINK_RE.lastIndex = 0;
	const m = LINK_RE.exec(input);
	if (!m) return false;
	return SAFE_SCHEME.test(m[2]);
}

const RICH_MARKDOWN_RE = /(^|\n)#{1,6}\s+\S|\*\*[^\s*][^*]*\*\*|(^|[\s])\*[^\s*][^*]*\*|(^|[\s])_[^\s_][^_]*_|`[^`]+`|(^|\n)([-*+]|\d+\.)\s+\S/;

export function hasMarkdownContent(input: string): boolean {
	if (!input) return false;
	if (hasMarkdownLink(input)) return true;
	return RICH_MARKDOWN_RE.test(input);
}

export type InlineSegment =
	| { type: 'text'; value: string }
	| { type: 'bold'; segments: InlineSegment[] }
	| { type: 'italic'; segments: InlineSegment[] }
	| { type: 'code'; value: string }
	| { type: 'link'; text: string; href: string };

export type MarkdownBlock =
	| { type: 'heading'; level: 1 | 2 | 3 | 4 | 5 | 6; segments: InlineSegment[] }
	| { type: 'paragraph'; segments: InlineSegment[] }
	| { type: 'list'; ordered: boolean; items: InlineSegment[][] };

const HEADING_RE = /^(#{1,6})\s+(.*)$/;
const UNORDERED_LIST_RE = /^[-*+]\s+(.*)$/;
const ORDERED_LIST_RE = /^\d+\.\s+(.*)$/;

export function parseInlineMarkdown(input: string): InlineSegment[] {
	if (!input) return [];
	const segments: InlineSegment[] = [];
	let i = 0;
	let textBuf = '';
	const flushText = () => {
		if (textBuf) {
			segments.push({ type: 'text', value: textBuf });
			textBuf = '';
		}
	};
	while (i < input.length) {
		const ch = input[i];
		if (ch === '`') {
			const end = input.indexOf('`', i + 1);
			if (end > i + 1) {
				flushText();
				segments.push({ type: 'code', value: input.slice(i + 1, end) });
				i = end + 1;
				continue;
			}
		}
		if (ch === '[') {
			LINK_RE.lastIndex = i;
			const m = LINK_RE.exec(input);
			if (m && m.index === i && SAFE_SCHEME.test(m[2])) {
				flushText();
				segments.push({ type: 'link', text: m[1], href: m[2] });
				i += m[0].length;
				continue;
			}
		}
		if (ch === '*' || ch === '_') {
			const isDouble = input[i + 1] === ch;
			const marker = isDouble ? ch + ch : ch;
			const start = i + marker.length;
			if (start < input.length && input[start] !== ' ') {
				const end = input.indexOf(marker, start);
				if (end > start) {
					const closingChar = input[end - 1];
					if (closingChar !== ' ' && (isDouble || input[end + 1] !== ch)) {
						flushText();
						const inner = parseInlineMarkdown(input.slice(start, end));
						segments.push({ type: isDouble ? 'bold' : 'italic', segments: inner });
						i = end + marker.length;
						continue;
					}
				}
			}
		}
		textBuf += ch;
		i++;
	}
	flushText();
	return segments;
}

export function parseMarkdownBlocks(input: string): MarkdownBlock[] {
	if (!input) return [];
	const lines = input.replace(/\r\n?/g, '\n').split('\n');
	const blocks: MarkdownBlock[] = [];
	let paragraphLines: string[] = [];
	let listItems: string[] | null = null;
	let listOrdered = false;

	const flushParagraph = () => {
		if (paragraphLines.length === 0) return;
		const text = paragraphLines.join('\n');
		blocks.push({ type: 'paragraph', segments: parseInlineMarkdown(text) });
		paragraphLines = [];
	};
	const flushList = () => {
		if (!listItems) return;
		blocks.push({
			type: 'list',
			ordered: listOrdered,
			items: listItems.map((item) => parseInlineMarkdown(item))
		});
		listItems = null;
	};

	for (const raw of lines) {
		const line = raw.trimEnd();
		if (line.trim() === '') {
			flushParagraph();
			flushList();
			continue;
		}
		const headingMatch = HEADING_RE.exec(line);
		if (headingMatch) {
			flushParagraph();
			flushList();
			const level = headingMatch[1].length as 1 | 2 | 3 | 4 | 5 | 6;
			blocks.push({ type: 'heading', level, segments: parseInlineMarkdown(headingMatch[2]) });
			continue;
		}
		const unorderedMatch = UNORDERED_LIST_RE.exec(line);
		const orderedMatch = !unorderedMatch ? ORDERED_LIST_RE.exec(line) : null;
		if (unorderedMatch || orderedMatch) {
			flushParagraph();
			const ordered = !!orderedMatch;
			if (listItems && listOrdered !== ordered) flushList();
			if (!listItems) {
				listItems = [];
				listOrdered = ordered;
			}
			listItems.push((unorderedMatch ?? orderedMatch)![1]);
			continue;
		}
		flushList();
		paragraphLines.push(line);
	}
	flushParagraph();
	flushList();
	return blocks;
}

export function stripMarkdownSyntax(input: string): string {
	if (!input) return '';
	const blocks = parseMarkdownBlocks(input);
	return blocks
		.map((block) => {
			if (block.type === 'list') {
				return block.items.map((item) => inlineToPlain(item)).join('\n');
			}
			return inlineToPlain(block.segments);
		})
		.join('\n')
		.replace(/\n{2,}/g, '\n')
		.trim();
}

function inlineToPlain(segments: InlineSegment[]): string {
	let out = '';
	for (const seg of segments) {
		if (seg.type === 'text') out += seg.value;
		else if (seg.type === 'code') out += seg.value;
		else if (seg.type === 'link') out += seg.text;
		else out += inlineToPlain(seg.segments);
	}
	return out;
}

File: android-native/app/src/main/kotlin/ru/tinyops/turboist/nativeapp/ui/markdown/Markdown.kt
package ru.tinyops.turboist.nativeapp.ui.markdown

/**
 * The subset of Markdown a task description is written in.
 *
 * It is deliberately the same subset the web client renders, parsed by the same
 * rules, because the two clients read one another's text: a description typed on
 * a phone is read in a browser the next morning and must look like what its
 * author meant. Anything outside the subset stays visible as the characters that
 * were typed rather than disappearing — a description is the user's own text,
 * and swallowing part of it is worse than showing a stray asterisk.
 *
 * Links are limited to schemes that cannot act on the device on their own. A
 * description arrives from the server and may have been written anywhere, so a
 * link is only turned into something tappable when it is plainly a web address,
 * a mail address, or a path inside the product.
 */
object Markdown {
    private val LINK = Regex("""\[([^\]]+)\]\(([^)\s]+)\)""")
    private val SAFE_SCHEME = Regex("""^(https?://|mailto:|/|#)""", RegexOption.IGNORE_CASE)
    private val HEADING = Regex("""^(#{1,6})\s+(.*)$""")
    private val UNORDERED_ITEM = Regex("""^[-*+]\s+(.*)$""")
    private val ORDERED_ITEM = Regex("""^\d+\.\s+(.*)$""")

    /**
     * Whether the text carries anything worth rendering as Markdown.
     *
     * Used to decide between showing a description as formatted text and showing
     * it as the plain text it was typed as. Plain prose is left alone: running it
     * through a renderer that changes nothing only costs the user the ability to
     * see their own spacing.
     */
    private val RICH =
        Regex(
            """(^|\n)#{1,6}\s+\S""" +
                """|\*\*[^\s*][^*]*\*\*""" +
                """|(^|\s)\*[^\s*][^*]*\*""" +
                """|(^|\s)_[^\s_][^_]*_""" +
                """|`[^`]+`""" +
                """|(^|\n)([-*+]|\d+\.)\s+\S""",
        )

    /** True when [input] holds a link this renderer would make tappable. */
    fun hasLink(input: String): Boolean {
        val match = LINK.find(input) ?: return false
        return SAFE_SCHEME.containsMatchIn(match.groupValues[2])
    }

    /** True when [input] holds any markup at all, a link included. */
    fun hasContent(input: String): Boolean {
        if (input.isEmpty()) return false
        return hasLink(input) || RICH.containsMatchIn(input)
    }

    /** The text cut into blocks: headings, paragraphs and lists, in the order written. */
    fun blocks(input: String): List<MarkdownBlock> {
        if (input.isEmpty()) return emptyList()
        val builder = BlockBuilder()
        for (raw in input.replace("\r\n", "\n").replace('\r', '\n').split('\n')) {
            builder.take(raw.trimEnd())
        }
        return builder.finish()
    }

    /**
     * One block of text cut into its runs: bold, italic, code, links and the
     * plain text between them.
     *
     * A marker with no partner, or one wrapped around nothing, is not markup and
     * is kept as the character it is.
     */
    fun inline(input: String): List<InlineSegment> {
        if (input.isEmpty()) return emptyList()
        val segments = mutableListOf<InlineSegment>()
        val text = StringBuilder()
        var i = 0

        fun flush() {
            if (text.isEmpty()) return
            segments += InlineSegment.Text(text.toString())
            text.clear()
        }

        while (i < input.length) {
            val consumed = readCode(input, i) ?: readLink(input, i) ?: readEmphasis(input, i)
            if (consumed == null) {
                text.append(input[i])
                i++
                continue
            }
            flush()
            segments += consumed.segment
            i = consumed.next
        }
        flush()
        return segments
    }

    /** A run of code between single backticks. */
    private fun readCode(
        input: String,
        at: Int,
    ): Consumed? {
        if (input[at] != '`') return null
        val end = input.indexOf('`', at + 1)
        if (end <= at + 1) return null
        return Consumed(InlineSegment.Code(input.substring(at + 1, end)), end + 1)
    }

    /** A `[text](target)` link, when the target is a scheme this renderer will open. */
    private fun readLink(
        input: String,
        at: Int,
    ): Consumed? {
        if (input[at] != '[') return null
        val match = LINK.matchAt(input, at) ?: return null
        val href = match.groupValues[2]
        if (!SAFE_SCHEME.containsMatchIn(href)) return null
        return Consumed(InlineSegment.Link(match.groupValues[1], href), at + match.value.length)
    }

    /**
     * A run wrapped in `*`/`_` for italic or `**`/`__` for bold.
     *
     * The marker has to hug its text on both sides, and a single marker must not
     * be the first half of a double one, which is what keeps `**bold**` from
     * being read as an empty italic followed by stray text.
     */
    private fun readEmphasis(
        input: String,
        at: Int,
    ): Consumed? {
        val marker = input[at]
        if (marker != '*' && marker != '_') return null
        val doubled = at + 1 < input.length && input[at + 1] == marker
        val width = if (doubled) 2 else 1
        val start = at + width
        if (start >= input.length || input[start] == ' ') return null
        val end = input.indexOf(if (doubled) "$marker$marker" else marker.toString(), start)
        if (end <= start) return null
        if (input[end - 1] == ' ') return null
        if (!doubled && end + 1 < input.length && input[end + 1] == marker) return null
        val inner = inline(input.substring(start, end))
        val segment = if (doubled) InlineSegment.Bold(inner) else InlineSegment.Italic(inner)
        return Consumed(segment, end + width)
    }

    private data class Consumed(
        val segment: InlineSegment,
        val next: Int,
    )

    /** Collects lines into blocks, closing whatever the current line ends. */
    private class BlockBuilder {
        private val blocks = mutableListOf<MarkdownBlock>()
        private val paragraph = mutableListOf<String>()
        private var items: MutableList<String>? = null
        private var ordered = false

        fun take(line: String) {
            if (line.isBlank()) {
                closeParagraph()
                closeList()
                return
            }
            val heading = HEADING.find(line)
            if (heading != null) {
                closeParagraph()
                closeList()
                blocks += MarkdownBlock.Heading(heading.groupValues[1].length, inline(heading.groupValues[2]))
                return
            }
            val unordered = UNORDERED_ITEM.find(line)
            val numbered = if (unordered == null) ORDERED_ITEM.find(line) else null
            if (unordered != null || numbered != null) {
                closeParagraph()
                val isOrdered = numbered != null
                if (items != null && ordered != isOrdered) closeList()
                if (items == null) {
                    items = mutableListOf()
                    ordered = isOrdered
                }
                items?.add((unordered ?: numbered)!!.groupValues[1])
                return
            }
            closeList()
            paragraph += line
        }

        fun finish(): List<MarkdownBlock> {
            closeParagraph()
            closeList()
            return blocks.toList()
        }

        private fun closeParagraph() {
            if (paragraph.isEmpty()) return
            blocks += MarkdownBlock.Paragraph(inline(paragraph.joinToString("\n")))
            paragraph.clear()
        }

        private fun closeList() {
            val open = items ?: return
            blocks += MarkdownBlock.Items(ordered, open.map(::inline))
            items = null
        }
    }
}

/** A run of text inside one block. */
sealed interface InlineSegment {
    data class Text(val value: String) : InlineSegment

    data class Bold(val segments: List<InlineSegment>) : InlineSegment

    data class Italic(val segments: List<InlineSegment>) : InlineSegment

    data class Code(val value: String) : InlineSegment

    data class Link(val text: String, val href: String) : InlineSegment
}

/** One block of a description. */
sealed interface MarkdownBlock {
    /** A heading. [level] is 1 to 6, as many as the hashes that introduced it. */
    data class Heading(val level: Int, val segments: List<InlineSegment>) : MarkdownBlock

    data class Paragraph(val segments: List<InlineSegment>) : MarkdownBlock

    /** A list. [ordered] tells a numbered list from a bulleted one. */
    data class Items(val ordered: Boolean, val items: List<List<InlineSegment>>) : MarkdownBlock
}
